sdk/go/common/resource/properties_diff.go

```diff
@@ -392,3 +392,271 @@ func (v PropertyValue) DeepEquals(other PropertyValue) bool {
+				Updates: updates,
+			},
+		}
+	}
+	if v.IsObject() && other.IsObject() {
+		old := v.ObjectValue()
+		new := other.ObjectValue()
+		if diff := old.DiffIncludeUnknowns(new, ignoreKeys...); diff != nil {
+			return &ValueDiff{
+				Old:    v,
+				New:    other,
+				Object: diff,
+			}
+		}
+		return nil
+	}
+
+	// If we got here, either the values are primitives, or they weren't the same type; do a simple diff.
+	if v.DeepEqualsIncludeUnknowns(other) {
+		return nil
+	}
+	return &ValueDiff{Old: v, New: other}
+}
+
+func (props PropertyMap) DeepEqualsIncludeUnknowns(other PropertyMap) bool {
+	// If any in props either doesn't exist, or is of a different value, return false.
+	for _, k := range props.StableKeys() {
+		v := props[k]
+		if p, has := other[k]; has {
+			if !v.DeepEqualsIncludeUnknowns(p) {
+				return false
+			}
+		} else if v.HasValue() && !v.IsComputed() {
+			return false
+		}
+	}
+
+	// If the other map has properties that this map doesn't have, return false.
+	for _, k := range other.StableKeys() {
+		if _, has := props[k]; !has && other[k].HasValue() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (v PropertyValue) DeepEqualsIncludeUnknowns(other PropertyValue) bool {
+	// Computed values are always equal.
+	if v.IsComputed() && other.IsComputed() {
+		return true
+	}
+
+	// Anything is equal to a computed
+	if v.IsComputed() || other.IsComputed() {
+		return true
+	}
+
+	// Arrays are equal if they are both of the same size and elements are deeply equal.
+	if v.IsArray() {
+		if !other.IsArray() {
+			return false
+		}
+		va := v.ArrayValue()
+		oa := other.ArrayValue()
+		if len(va) != len(oa) {
+			return false
+		}
+		for i, elem := range va {
+			if !elem.DeepEqualsIncludeUnknowns(oa[i]) {
+				return false
+			}
+		}
+		return true
+	}
+
+	// Assets and archives enjoy value equality.
+	if v.IsAsset() {
+		if !other.IsAsset() {
+			return false
+		}
+		return v.AssetValue().Equals(other.AssetValue())
+	} else if v.IsArchive() {
+		if !other.IsArchive() {
+			return false
+		}
+		return v.ArchiveValue().Equals(other.ArchiveValue())
+	}
+
+	// Object values are equal if their contents are deeply equal.
+	if v.IsObject() {
+		if !other.IsObject() {
+			return false
+		}
+		vo := v.ObjectValue()
+		oa := other.ObjectValue()
+		return vo.DeepEqualsIncludeUnknowns(oa)
+	}
+
+	// Secret are equal if the value they wrap are equal.
+	if v.IsSecret() {
+		if !other.IsSecret() {
+			return false
+		}
+		vs := v.SecretValue()
+		os := other.SecretValue()
+
+		return vs.Element.DeepEqualsIncludeUnknowns(os.Element)
+	}
+
+	// Resource references are equal if they refer to the same resource. The package version is ignored.
+	if v.IsResourceReference() {
+		if !other.IsResourceReference() {
+			return false
+		}
+		vr := v.ResourceReferenceValue()
+		or := other.ResourceReferenceValue()
+
+		if vr.URN != or.URN {
+			return false
+		}
+
+		vid, oid := vr.ID, or.ID
+		if vid.IsComputed() && oid.IsComputed() {
```

Comment:
&& or || here?

Reply:
I think `||` actually. There's some redundancy above with IsComputed checks at the top of this function doing `&&` and `||`